api/grpcserver/globalstate_service.go

```diff
@@ -0,0 +1,457 @@
+
+	return status.Errorf(codes.Unimplemented, "this endpoint has not yet been implemented")
+}
+
+// AppEventStream exposes a stream of emitted app events
+func (s GlobalStateService) AppEventStream(request *pb.AppEventStreamRequest, stream pb.GlobalStateService_AppEventStreamServer) error {
+	log.Info("GRPC GlobalStateService.AppEventStream")
+
+	// TODO: implement me! We don't currently have any app events
+	// See https://github.com/spacemeshos/go-spacemesh/issues/2074
+
+	return status.Errorf(codes.Unimplemented, "this endpoint has not yet been implemented")
+}
+
+// GlobalStateStream exposes a stream of global data data items: rewards, receipts, account info, global state hash
+func (s GlobalStateService) GlobalStateStream(in *pb.GlobalStateStreamRequest, stream pb.GlobalStateService_GlobalStateStreamServer) error {
+	log.Info("GRPC GlobalStateService.GlobalStateStream")
+
+	if in.GlobalStateDataFlags == uint32(pb.GlobalStateDataFlag_GLOBAL_STATE_DATA_FLAG_UNSPECIFIED) {
+		return status.Errorf(codes.InvalidArgument, "`GlobalStateDataFlags` must set at least one bitfield")
+	}
+
+	filterAccount := in.GlobalStateDataFlags&uint32(pb.GlobalStateDataFlag_GLOBAL_STATE_DATA_FLAG_ACCOUNT) != 0
+	filterReward := in.GlobalStateDataFlags&uint32(pb.GlobalStateDataFlag_GLOBAL_STATE_DATA_FLAG_REWARD) != 0
+	filterReceipt := in.GlobalStateDataFlags&uint32(pb.GlobalStateDataFlag_GLOBAL_STATE_DATA_FLAG_TRANSACTION_RECEIPT) != 0
+	filterState := in.GlobalStateDataFlags&uint32(pb.GlobalStateDataFlag_GLOBAL_STATE_DATA_FLAG_GLOBAL_STATE_HASH) != 0
+
+	// Subscribe to the various streams
+	var (
+		channelAccount chan types.Address
+		channelReward  chan events.Reward
+		channelReceipt chan events.TxReceipt
+		channelLayer   chan events.NewLayer
+	)
+	if filterAccount {
+		channelAccount = events.GetAccountChannel()
+	}
+	if filterReward {
+		channelReward = events.GetRewardChannel()
+	}
+	if filterReceipt {
+		channelReceipt = events.GetReceiptChannel()
+	}
+	if filterState {
+		// Whenever new state is applied to the mesh, a new layer is reported.
+		// There is no separate reporting specifically for new state.
+		channelLayer = events.GetLayerChannel()
+	}
+
+	for {
+		select {
+		case updatedAccount, ok := <-channelAccount:
+			if !ok {
+				// we could handle this more gracefully, by no longer listening
+				// to this stream but continuing to listen to the other stream,
+				// but in practice one should never be closed while the other is
+				// still running, so it doesn't matter
+				log.Info("account channel closed, shutting down")
+				return nil
+			}
+			// The Reporter service just sends us the account address. We are responsible
+			// for looking up the other required data here. Get the account balance and
+			// nonce.
+			balance := s.Mesh.GetBalance(updatedAccount)
+			counter := s.Mesh.GetNonce(updatedAccount)
+			if err := stream.Send(&pb.GlobalStateStreamResponse{Datum: &pb.GlobalStateData{Datum: &pb.GlobalStateData_AccountWrapper{
```

Comment:
I'd create a constructor method for these PB types for bette readability

Reply:
Good point. I agree that would be better. There are dozens of them, and that's a pretty big project. Are you okay if I open another task for this and leave it as an improvement?

Reply:
#2118